### PR TITLE
Update robust loss

### PR DESCRIPTION
**Pull request by xphsean12**
## Motivation and Context
Current Robust Loss cannot deal with outlinear existing in a optimization problem

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: theseus/core/robust_cost_function.py [hunk 1]
```diff
@@ -80,20 +82,33 @@ def error(self) -> torch.Tensor:
 
     def weighted_error(self) -> torch.Tensor:
         weighted_error = self.cost_function.weighted_error()
-        squared_norm = torch.sum(weighted_error**2, dim=1, keepdim=True)
-        error_loss = self.loss.evaluate(squared_norm, self.log_loss_radius.tensor)
-
+        
+        if self.RobustSum:
```
Comment (luisenp) [inline theseus/core/robust_cost_function.py:86]: I think this can be done much more simply. If I'm not mistaken, you can keep all of the old code and just add the following lines at the beginning of `weighted_error()`
```python
if self.flatten_dims:
    weighted_error = weighted_error.view(-1, 1)
```
and then reshape the output back to `(-1, self.dim())` before returning. 

A similar procedure can be followed for `weighted_jacobians`.

File: theseus/core/robust_cost_function.py [hunk 2]
```diff
@@ -80,20 +82,33 @@ def error(self) -> torch.Tensor:
...
+            torch.ones_like(weighted_error)
+            * (error_loss / self.dim() + RobustCostFunction._EPS).sqrt()
+            )
+        else:
+            squared_norm = weighted_error**2
+            error_loss = self.loss.evaluate(squared_norm, self.log_loss_radius.tensor)
+            return (
+              (error_loss  + RobustCostFunction._EPS).sqrt()
+            )
+        # Inside will compare squared_norm with exp(radius)
```
Comment (luisenp) [inline theseus/core/robust_cost_function.py:99]: Please remove any debug comments.

File: theseus/core/robust_loss.py [hunk 3]
```diff
@@ -11,13 +11,14 @@
 
 
 class RobustLoss(abc.ABC):
+    # .suqre() make it easier to compare with residual.square(), which is x
     @classmethod
     def evaluate(cls, x: torch.Tensor, log_radius: torch.Tensor) -> torch.Tensor:
-        return cls._evaluate_impl(x, log_radius.exp())
-
+        return cls._evaluate_impl(x, log_radius.square())
```
Comment (luisenp) [inline theseus/core/robust_loss.py:17]: We should keep this as `exp()`.

File: theseus/core/robust_loss.py [hunk 4]
```diff
@@ -49,4 +50,4 @@ def _evaluate_impl(x: torch.Tensor, radius: torch.Tensor) -> torch.Tensor:
 
     @staticmethod
     def _linearize_impl(x: torch.Tensor, radius: torch.Tensor) -> torch.Tensor:
-        return torch.sqrt(radius / torch.max(x, radius) + _LOSS_EPS)
+        return torch.sqrt(radius / torch.max(x, radius) + _LOSS_EPS)
```
Comment (luisenp) [inline theseus/core/robust_loss.py:53]: Missing line at end of file.

File: tutorials/test_RobustLoss.ipynb [hunk 5]
```diff
@@ -0,0 +1,731 @@
+{
```
Comment (luisenp) [inline tutorials/test_RobustLoss.ipynb:1]: For the unit test, the right place to do it is [this](https://github.com/facebookresearch/theseus/blob/main/tests/core/test_robust_cost.py) file. There is no need to write a full example. We only need two versions of a small objective and compare that the both versions lead to the same error vectors. We can remove this notebook.

File: theseus/core/robust_cost_function.py [hunk 6]
```diff
@@ -107,15 +122,43 @@ def weighted_jacobians_error(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
...
-        squared_norm = torch.sum(weighted_error**2, dim=1, keepdim=True)
-        rescale = (
-            self.loss.linearize(squared_norm, self.log_loss_radius.tensor)
-            + RobustCostFunction._EPS
-        ).sqrt()
-
-        return [
-            rescale.view(-1, 1, 1) * jacobian for jacobian in weighted_jacobians
-        ], rescale * weighted_error
+        #squared_norm = torch.sum(weighted_error**2, dim=1, keepdim=True)
```
Comment (luisenp) [inline theseus/core/robust_cost_function.py:125]: See comment above on a suggestion for implementing this more easily.

**Review by luisenp — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution! I left some comments that need to be addressed. 

Also CI is failing due to some lint errors. Can you please install `pip install -r requirements/dev.txt` and also [pre-commit hooks](https://pre-commit.com/#usage), so that they take care of lint errors automatically. 

Thanks!

File: theseus/core/robust_cost_function.py (hunk 1 above)
Comment (xphsean12) [reply theseus/core/robust_cost_function.py:86]: I am not sure what you mean. And why I need to reshape it. Because what I want is given any shape of error, like (Batch, Dim) and the corresponding radius with the shape of (Batch, 1), they will all be passed to the evaluation function self.loss.evaluate and return the scaled loss. 

image

File: theseus/core/robust_loss.py (hunk 3 above)
Comment (xphsean12) [reply theseus/core/robust_loss.py:17]: Sure. Could I know the initiative behind it?

File: tutorials/test_RobustLoss.ipynb (hunk 5 above)
Comment (xphsean12) [reply tutorials/test_RobustLoss.ipynb:1]: I use notebook to test to avoid the old version will mess up with the new one. Because I have installed Theseus 0.1.4 and do not want to make any changes to it. And any of the function that call th.xx will call the original Theseus function. Do you have any good ideas? Or I can just replace my new script file with the one in the environment site-package?

File: theseus/core/robust_cost_function.py (hunk 1 above)
Comment (luisenp) [reply theseus/core/robust_cost_function.py:86]: It's a bit hard to explain, but it works, I just tested it. I'll put another PR soon.

Comment (luisenp) [reply theseus/core/robust_cost_function.py:86]: Please take a look at #503. The code for the jacobians part was a bit tricker than I suggested above, but I was still able to mostly use all of the old computation. Also look at the unit test I added and let me know if covers the expected behavior. Thanks!

File: theseus/core/robust_loss.py (hunk 3 above)
Comment (fantaosha) [reply theseus/core/robust_loss.py:17]: Since square() is not monotonous, I personally prefer 'exp()', which is more useful when we try to learn a robust loss radius.

**Review by fantaosha — COMMENTED** [review]
Add some more comments in #503.

File: theseus/core/robust_loss.py (hunk 3 above)
Comment (xphsean12) [reply theseus/core/robust_loss.py:17]: > 

Oh! That's great! Because at a very early time, I tried to let the system automatically learn the Radius, and I failed at that time (performance was not good). But with the new version, the training outcome may change.

File: theseus/core/robust_cost_function.py (hunk 1 above)
Comment (xphsean12) [reply theseus/core/robust_cost_function.py:86]: > It's a bit hard to explain, but it works, I just tested it. I'll put another PR soon.

Oh! I see what you mean! You treat every element of the original cost as the original "square norm"
